src/clj/ataru/haku/haku_service.clj

```diff
@@ -110,35 +107,59 @@
          (map remove-organization-oid)
          (util/group-by-first :key))))
 
+(defn- get-hakuKohteet-without-selection
+  [tarjonta-haut
+   tarjonta-service]
+  (->> (keys tarjonta-haut)
+       (mapcat #(tarjonta/hakukohde-search
+                  tarjonta-service
+                  %
+                  nil)) tarjonta-haut))
+
+(defn- add-selection-to-hakukohteet
+  [hakukohteet-without-selection]
+    (let [hakukohdeoids (map #(:oid %) hakukohteet-without-selection)
+          hakukohde-oids-with-selection-state-used (hakukohde-store/selection-state-used-in-hakukohdes? hakukohdeoids)]
+        (->> hakukohteet-without-selection
+             (map (fn [{hakukohde-oid :oid :as hakukohde}]
+                    (assoc
+                      hakukohde
+                      :selection-state-used (some hakukohde-oid hakukohde-oids-with-selection-state-used))))
+              (util/group-by-first :oid))))
+
+
 (defn get-haut
   [ohjausparametrit-service
    organization-service
    tarjonta-service
    get-haut-cache
    session
    show-hakukierros-paattynyt?]
-  (let [tarjonta-haut (get-tarjonta-haut ohjausparametrit-service
-                                         organization-service
-                                         tarjonta-service
-                                         get-haut-cache
-                                         session
-                                         show-hakukierros-paattynyt?)]
-    {:tarjonta-haut    tarjonta-haut
-     :direct-form-haut (get-direct-form-haut organization-service get-haut-cache session)
-     :haut             (->> (keys tarjonta-haut)
-                            (keep #(tarjonta/get-haku tarjonta-service %))
-                            (util/group-by-first :oid))
-     :hakukohteet      (->> (keys tarjonta-haut)
-                            (mapcat #(tarjonta/hakukohde-search
-                                       tarjonta-service
-                                       %
-                                       nil))
-                            (map (fn [{hakukohde-oid :oid :as hakukohde}]
-                                   (assoc
-                                     hakukohde
-                                     :selection-state-used
-                                     (hakukohde-store/selection-state-used-in-hakukohde? hakukohde-oid))))
-                            (util/group-by-first :oid))
-     :hakukohderyhmat  (util/group-by-first
-                         :oid
-                         (filter :active? (organization-service/get-hakukohde-groups organization-service)))}))
+  (let [startedHaku (System/currentTimeMillis)]
+    (log/debug (str "!!!!Started haku at " (quot startedHaku 1000) " s"))
```

Comment:
Muutin nämä debugiksi tähän, mut mietin tarviiko näitä loggeja ylipäätänsä jättää tänne.

Voivat auttaa ongelman selvittämisessä, mutta olisihan tämä vähän siistimpi ilman niitä.

Reply:
Tuo let-pyramidi ei kyllä ole kovin kaunis rakenne muutenkaan, houkuttaisi päästä siitä eroon. :) Saisikohan tuon pienellä vaivalla muokattua niin, että kaikki haut tehtäisiin yhdessä letissä mutta yksitellen kääräistyinä jonkinlaiseen simppeliin ajanottofunktioon, joka logittaisi jotain jos siinä kestää pidempään kuin joku järkevä kynnysarvo.

Reply:
Se etu tuossa ainankin olisi, että voisi virittää logituksen hiljaiseksi nykykestolla, mutta saisi vaikka vähän WARN-mökää logille jos kestot nykyisestä pitenevät.

Reply:
Olisiko se jotain 7 sekunttia? Näyttäisi hahtuvalla kestävän noin 4 sekkaa nyt,

Tai 10.

Reply:
Kuulostaisi järkevältä, nykyinen keskivertokesto*1.5 esim.